File: Leetcode/LinkedList/ReverseLinkedList/ReverseLinkedList.py
#  206. Reverse Linked List
#  https://leetcode.com/problems/reverse-linked-list/

# Definition for singly-linked list.
from typing import Counter, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __str__(self):
        temp = f"{self.val}"
        nextNode = self.next
        while nextNode is not None:
            temp += f", {nextNode.val}"
            nextNode = nextNode.next
        return temp


# recursively
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self._reverseList(head, None)

    def _reverseList(self, head: Optional[ListNode], prev: Optional[ListNode]):
        if head is None:
            return prev

        next = head.next
        head.next = prev
        return self._reverseList(next, head)


# iteratively
class Solution1:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None

        while head:
            curr = head
            head = head.next
            curr.next = prev
            prev = curr

        return prev


if __name__ == "__main__":
    sol = Solution()

    head = ListNode(1)
    nextNode = head
    for i in range(2, 6):
        temp = ListNode(i)
        nextNode.next = temp
        nextNode = nextNode.next

    print(f"head: {head}")

    result: ListNode = sol.reverseList(head)
    print(f"result: {result}")
    # result: ListNode = sol.reverseList(ListNode())
    # print(f"result: {result}")

File: Leetcode/LinkedList/ReverseLinkedList/ReverseLinkedList.java
package LinkedList.ReverseLinkedList;
// 206. Reverse Linked List
// https://leetcode.com/problems/reverse-linked-list/

import java.util.LinkedList;
import java.util.List;

// * Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;

    ListNode() {}

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    @Override
    public String toString() {
        String temp = String.valueOf(this.val);
        ListNode next = this.next;
        while (next != null) {
            temp += ", " + next.val;
            next = next.next;
        }
        return temp;
    }
}


// O(n)
class Solution {
    public ListNode reverseList(ListNode head) {
        return _reverseList(head, null);
    }

    private ListNode _reverseList(ListNode head, ListNode prev) {
        if (head == null) {
            return prev;
        }

        ListNode next = head.next;
        head.next = prev;

        return _reverseList(next, head);
    }
}


// O(n)
class Solution2 {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode curr = head;
            head = head.next;
            curr.next = prev;
            prev = curr;
        }

        return prev;
    }
}


// O(n)
class Solution1 {
    public ListNode reverseList(ListNode head) {
        if (head == null) {
            return null;
        }

        ListNode result = new ListNode();

        ListNode it = head;
        while (it.next != null) {
            ListNode temp = new ListNode(it.val);

            temp.next = result.next;
            result.next = temp;

            it = it.next;
        }
        result.val = it.val;

        return result;
    }
}


public class ReverseLinkedList {
    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        ListNode head = new ListNode(1);
        ListNode next = head;
        while (next.next != null)
            next = next.next;
        for (int i = 2; i <= 5; ++i) {
            ListNode temp = new ListNode(i);
            next.next = temp;
            next = next.next;
        }

        System.out.println(head);

        Solution sol = new Solution();

        System.out.println(sol.reverseList(head));
    }

}
